multi-layer-app-config/README.md

```diff
@@ -0,0 +1,193 @@
+
+## User stories
+
+1. As a GitOps user, I want to have as many base layers in my GitOps repo as I want. On each layer, I want to be able to
+   override App config with selected keys only and not have to provide a full configuration.
+1. As a GitOps user, I want to be able to use setups that come from multiple base layers at once. As an example, I want
+   to have a `dev/stage/prod` base that sets app's resources based on deployment stage (by overriding a single
+   config key) and I want to have regional bases like `east/west` that configure app's allowed IP ranges (again, by
+   overriding another single config property).
+
+## Possible solutions
+
+### Enhancing App CR
+
+We can extend App CR with on optional list of `ConfigMap`/`Secret` objects. Configuration coming from `config:` and
+`userConfig:` properties (if given) is applied at the end of this list (to maintain backward compatibility).
+`app-operator` does merging of all layers on the list, from top to bottom, instead of just merging the two properties.
+
+The idea is modeled after Flux's [HelmRelease configuration](https://fluxcd.io/docs/components/helm/api/#helm.toolkit.fluxcd.io/v2beta1.ValuesReference).
+
+#### Example
+
+With empty config list, works as it was so far:
+
+```yaml
+apiVersion: application.giantswarm.io/v1alpha1
+kind: App
+spec:
+  catalog: giantswarm
+  config:
+    configMap:
+      name: ingress-controller-values
+      namespace: m2m01
+  configs: []  # <-- new
+  name: nginx-ingress-controller-app
+  namespace: kube-system
+  userConfig:
+    configMap:
+      name: nginx-ingress-controller-app-user-values
+      namespace: m2m01
+  version: 2.7.0
+```
+
+Using the new list instead of `config` and `userConfig`, works exactly the same as the one above:
+
+```yaml
+apiVersion: application.giantswarm.io/v1alpha1
+kind: App
+spec:
+  catalog: giantswarm
+  configs:  # <-- new
+    - kind: configMap
+      name: ingress-controller-values
+      namespace: m2m01
+    - kind: configMap
+      name: nginx-ingress-controller-app-user-values
+      namespace: m2m01
+  name: nginx-ingress-controller-app
+  namespace: kube-system
+  version: 2.7.0
+```
+
+Using the new list only and referencing more than 2 objects:
+
+```yaml
+apiVersion: application.giantswarm.io/v1alpha1
+kind: App
+spec:
+  catalog: giantswarm
+  configs:  # <-- new
+    - kind: configMap
+      name: ingress-controller-values
+      namespace: m2m01
+    - kind: configMap
+      name: nginx-ingress-controller-app-user-values
+      namespace: m2m01
+    - kind: secret
+      name: nginx-ingress-controller-admin-login
+      namespace: m2m01
+    - kind: configMap
+      name: nginx-ingress-controller-admin-account
+      namespace: m2m01
+  name: nginx-ingress-controller-app
+  namespace: kube-system
+  version: 2.7.0
+```
+
+#### Pros
+
+- solves the problem everywhere, in App CR and in GitOps
+- backward compatible
+- easy to implement
```

Comment:
But is it easy to release? Doesn't it tie back to the problem of patching release CRs? I mean, if I am not mistaken, implementation should be farily easy for unique App Operator and CAPx (org-namespaced) App Operators, because there are either latest or close to it, but there are also older versions of App Operator in use that we would need to patch in order to support these, right? Of course if customers upgrade the clusters then that's not a problem anymore.

Reply:
I think the benefit here is, that we only add a field. Adding a field in a CRD is a non-breaking change and could be rolled out easily. We acquire slightly more technical debt but the general roll-out should be no problem.

Reply:
I agree. This new version of app-op can just create a new release that customers have to upgrade to if they want to get benefits. We don't have to patch existing releases.